### PR TITLE
fix(jsonld): allow @id, @context and @type on denormalization 2

### PR DESCRIPTION
**Pull request by ili101**
| Q             | A
| ------------- | ---
| Branch?       | 3.3  
| Tickets       | Closes https://github.com/api-platform/core/issues/6225 
| License       | MIT
| Doc PR        | Feature already documented

Allaw `@id`, `@type`, `@context` with JsonLd. With this it's possible to use the `@id` as [described in the documentation ](https://api-platform.com/docs/core/serialization/#denormalization) while using `allow_extra_attributes: false`.
More info in https://github.com/api-platform/core/pull/6402.

### CODE REVIEW

**Comment by soyuka** [comment]
this is perfect, would you be able to add a test?

**Review by soyuka — APPROVED** [review]

File: features/main/standard_put.feature [hunk 1]
```diff
@@ -26,6 +26,32 @@ Feature: Spec-compliant PUT support
     }
     """
 
+  Scenario: Create a new resource with json-ld attributes
```
Comment (dunglas) [inline features/main/standard_put.feature:29]:
```suggestion
  Scenario: Create a new resource with JSON-LD attributes
```

nit

File: features/main/standard_put.feature [hunk 2]
```diff
@@ -26,6 +26,32 @@ Feature: Spec-compliant PUT support
...
+    {
+      "@context": "/contexts/StandardPut",
+      "@id": "/standard_puts/6",
+      "@type": "StandardPut",
+      "id": 6,
+      "foo": "a",
+      "bar": "b"
+    }
+    """
+
```
Comment (dunglas) [inline features/main/standard_put.feature:54]: Could you also add a test that ensures an error is thrown if the value of the `@id` key doesn't match the URL?

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 3]
```diff
@@ -148,9 +150,26 @@ public function denormalize(mixed $data, string $class, ?string $format = null,
...
         }
 
         return parent::denormalize($data, $class, $format, $context);
     }
+
+    protected function getAllowedAttributes(string|object $classOrObject, array $context, bool $attributesAsString = false): array|bool
+    {
+        $allowedAttributes = parent::getAllowedAttributes($classOrObject, $context, $attributesAsString);
+        if (\is_array($allowedAttributes) && ($context['api_denormalize'] ?? false)) {
+            $allowedAttributes = array_merge($allowedAttributes, ['@id', '@type', '@context']);
```
Comment (dunglas) [inline src/JsonLd/Serializer/ItemNormalizer.php:170]: Maybe should we allow all special keys (`@set`, `@container`, `@version`, `@nest` etc)?

**Comment by ili101** [comment]
> this is perfect, would you be able to add a test?

I was not available but I see a test was already added.
If it's useful in any way here is a [PHPUnit test](https://github.com/ili101/api-platform/blob/TestId/api/tests/Api/MainsTest.php) ([commit](https://github.com/ili101/api-platform/commit/c19da417399189b1524c29d380715472401fb8ea)) of the full usecase of `writableLink` using `@id` while `allow_extra_attributes: false`.

Side note I see you were discussing:
> Could you also add a test that ensures an error is thrown if the value of the `@id` key doesn't match the URL?

If the [`@id`](https://github.com/ili101/api-platform/blob/c19da417399189b1524c29d380715472401fb8ea/api/tests/Api/MainsTest.php#L41) line on the "Main" object is removed here **or** id is different from the URL, the "subs" array embedded objects will behave unexpectedly (removed or throw a duplicate error). Not sure if a problem or just needs to be documented as the description in the documentation only mentions the need/behavior for the `@id` on the sub embedded objects and not their parent too.

File: src/JsonLd/Serializer/ItemNormalizer.php (hunk 3 above)
Comment (soyuka) [reply src/JsonLd/Serializer/ItemNormalizer.php:170]: done I added all the keywords

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 4]
```diff
@@ -47,6 +49,29 @@ final class ItemNormalizer extends AbstractItemNormalizer
     use JsonLdContextTrait;
 
     public const FORMAT = 'jsonld';
+    public const JSONLD_KEYWORDS = [
```
Comment (dunglas) [inline src/JsonLd/Serializer/ItemNormalizer.php:52]: This could maybe be an `@internal` enum?

Comment (soyuka) [reply src/JsonLd/Serializer/ItemNormalizer.php:52]: too much trouble transforming an enum to an array this is more efficient imo, I could set `private` visibility?

File: src/Symfony/Routing/IriConverter.php [hunk 5]
```diff
@@ -78,6 +78,16 @@ public function getResourceFromIri(string $iri, array $context = [], ?Operation
             throw new InvalidArgumentException(sprintf('No resource associated to "%s".', $iri));
         }
 
+        foreach ($context['uri_variables'] ?? [] as $key => $value) {
+            if (!isset($parameters[$key]) || $parameters[$key] !== (string) $value) {
+                throw new InvalidArgumentException(sprintf('The iri "%s" does not reference the correct resource.', $iri));
+            }
+        }
+
+        if ($operation && $operation->getClass() !== $parameters['_api_resource_class']) {
```
Comment (soyuka) [inline src/Symfony/Routing/IriConverter.php:87]: use is_a to make this work with abstract classes

**Review by dunglas — APPROVED** [review]

**Comment by soyuka** [comment]
thanks @ili101 !

**Pull request merged**